onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -107,10 +107,21 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
     if (!p.second)
       continue;
 
-    if (provider_nodes.find(cur) == provider_nodes.end())
+    auto* node = graph.GetNode(cur);
+    if (provider_nodes.find(cur) == provider_nodes.end()) {
+      // Nodes not in provider_nodes are either have EP assigned or no kernel found on target EP.
+      // we assume these nodes will fallback to CPU, so add all direct consumers of all outputs to candidates.
```

Comment:
I'm not sure that we can generally assume that an unassigned node will be assigned to the CPU EP if there are other EPs which may be able to take the node.

Also, FYI I think another PR is trying to address a similar issue:
https://github.com/microsoft/onnxruntime/pull/13442#discussion_r1006138231

Reply:
If we don't have this assumption, then our current fallback design is wrong. Take CUDA EP as example, we use GetCpuPreferredNodes to get nodes might be in CPU, but when it goes back to GetCapability in CUDA EP, it just ignores those Ops from tentative nodes, assign CUDA EP to the others, but still leaves nodes from GetCpuPreferredNodes un-assigned. This means that other EP may also take them instead of CPU EP at the end. In real world, EPs other than CPU EP will be handled before CPU EP, and all of them should also call such fallback logic with same code, so these un-assigned nodes will keep un-assigned until it goes to CPU EP. The fix in this PR is just to keep such kernel-not-found nodes as un-assigned, by also make some nodes after that also keep un-assigned and relies on other following EPs to handle them, just as same logic before the fix.

The requires_matching_kernel flag added in PR #13442 has a big problem (at least for CUDA, not sure for DML). Take Gather->ReduceProd->Concat as example, here we don't have Shape before Gather, and ReduceProd has CPU kernel only. After change from #13442, since output of ReduceProd is added to cpu_output_args, then Concat will also be put to CPU (and maybe some other following nodes). This is unexpected because the ideal why is just keep ReduceProd in CPU. I.E., before #13442, it should be Gather->MemcpyToHost->ReduceProd->MemcpyFromHost->Concat, but after #13442, i will be Gather->MemcpyToHost->ReduceProd->Concat->...->MemcpyFromHost, it moves more nodes to CPU which will harm the perf.

Reply:
I also left a comment to #13442, but I saw that PR was already merged.

Reply:
Thanks for warning me on the other PR. I also noticed some slowdowns so I made a PR to revert it: https://github.com/microsoft/onnxruntime/pull/13605